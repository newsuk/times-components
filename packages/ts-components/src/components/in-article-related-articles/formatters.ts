import get from 'lodash.get';

export const formatRelatedArticles = (relatedArticleSlice: any) =>
  relatedArticleSlice.items.map(({ article }: any) => ({
    label: article.label,
    headline: article.shortHeadline || article.headline,
    link: article.url,
    summary:
      get(article, 'summary225[0].children[0].attributes.value') ||
      get(article, 'summary225[0].children[0].children[0].attributes.value'),
    publishedTime: article.publishedTime,
    image:
      get(article, 'leadAsset.crop169.url') ||
      get(article, 'leadAsset.posterImage.crop169.url'),
    byline: get(article, 'bylines[0].byline[0].children[0].attributes.value')
  }));
