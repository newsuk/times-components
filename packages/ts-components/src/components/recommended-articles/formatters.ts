import { SliceArticle } from '@times-components/ts-slices';

// TYPES

type Byline =
  | { __typename: string; type: 'author'; name: string }
  | { __typename: string; type: 'inline'; value: string };

type MediaCrop = {
  __typename: string;
  url: string;
  alt?: string;
  aspectRatio: string;
};
type Media = { __typename: string; crops: MediaCrop[] };

type SummaryText = { __typename: string; text: string };
type Summary = { __typename: string; children: SummaryText[] };

type Article = {
  __typename: string;
  url: string;
  slug: string;
  label: string;
  headline: string;
  publishedDateTime: string;
  bylines?: Byline[];
  summary?: Summary;
  media?: Media;
};

// HELPERS

const getBylines = (bylines?: Byline[]) =>
  bylines
    ? bylines
        .map(byline => (byline.type === 'author' ? byline.name : byline.value))
        .join('')
    : undefined;

const getSummary = (summary?: Summary) =>
  summary && summary.children
    ? summary.children.map(child => child.text).join('') + '...'
    : undefined;

const getImageCrops = (media?: Media) =>
  media && media.crops
    ? media.crops.map(crop => ({ url: crop.url, ratio: crop.aspectRatio }))
    : [];

// MAIN

const formatArticle = (article: Article): SliceArticle => {
  return {
    url: article.url,
    label: article.label,
    byline: getBylines(article.bylines),
    headline: article.headline,
    summary: getSummary(article.summary),
    datePublished: article.publishedDateTime,
    images: {
      alt: article.headline,
      crops: getImageCrops(article.media)
    }
  };
};

const getSliceName = (numOfArticles: number) => {
  switch (numOfArticles) {
    case 1:
      return 'RELATED_ARTICLE_1';
    case 2:
      return 'RELATED_ARTICLE_2';
    default:
      return 'RELATED_ARTICLE_3';
  }
};

export const getRecommendedArticlesSlice = (articles: Article[]) => {
  return {
    name: getSliceName(articles.length),
    children: articles
      .map((article: Article) => ({
        article: formatArticle(article)
      }))
      .slice(0, 3)
  };
};
