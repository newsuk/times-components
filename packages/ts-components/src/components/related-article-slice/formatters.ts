import { SliceArticle } from '@times-components/ts-slices';
import get from 'lodash.get';
import { Asset, RelatedArticle, RelatedArticles } from './types';

const decodeEntities = (str: string) =>
  str
    .replace(/&#(\d+);/g, (__: string, d: number) => String.fromCharCode(d))
    .replace(/\s+/g, ' ');

const isLabelEqualTo = (label: string, match?: string) =>
  label && match && label.toLowerCase() === match.toLowerCase();

const isLabelAttrValid = (title: string, label: string, byline?: string) =>
  !isLabelEqualTo(label, title) && !isLabelEqualTo(label, byline);

const getLabelAttributes = (label?: string) =>
  label ? label.split(' | ') : [];

const formatLabel = (title: string, article: Partial<SliceArticle>) =>
  getLabelAttributes(article.label)
    .map((a: string) => decodeEntities(a))
    .filter((a: string) => isLabelAttrValid(title, a, article.byline))
    .join(' | ');

const flattenAssets = (asset: Asset) => {
  const crops = [];
  if (asset.crop169) {
    crops.push({
      url: asset.crop169.url,
      ratio: '16:9'
    });
  }
  if (asset.crop32) {
    crops.push({
      url: asset.crop32.url,
      ratio: '3:2'
    });
  }
  return crops;
};
const formatArticle = (
  article: RelatedArticle,
  numberOfArticles: number
): SliceArticle => {
  const leadAssetCrops =
    (article.leadAsset &&
      article.leadAsset.posterImage &&
      flattenAssets(article.leadAsset.posterImage)) ||
    [];

  const leadAsset = {
    title: get(article, 'leadAsset.title'),
    crops: leadAssetCrops
  };
  const summary =
    numberOfArticles === 1 &&
    get(article, 'summary225[0].children[0].attributes.value');

  const byline = get(
    article,
    'bylines[0].byline[0].children[0].attributes.value'
  );

  const crops = (article.leadAsset && flattenAssets(article.leadAsset)) || [];

  return {
    id: article.id,
    url: article.url,
    label: formatLabel(leadAsset.title!, { label: article.label!, byline }),
    byline,
    headline: article.headline,
    summary,
    datePublished: article.publishedTime,
    dateUpdated: article.updatedTime,
    images: {
      alt: '',
      crops: [...crops, ...leadAssetCrops]
    }
  };
};

const getSliceName = (relatedArticleSlice: RelatedArticles) => {
  if (relatedArticleSlice.items.length === 1) {
    return 'RELATED_ARTICLE_1';
  }
  if (relatedArticleSlice.items.length === 2) {
    return 'LEAD_2';
  }
  return 'LEAD_1_AND_2';
};
const formatSlice = (relatedArticleSlice: RelatedArticles) => {
  return {
    name: getSliceName(relatedArticleSlice),
    children: relatedArticleSlice.items.map(relatedArticle => ({
      article: formatArticle(
        relatedArticle.article,
        relatedArticleSlice.items.length
      )
    }))
  };
};
export default formatSlice;
