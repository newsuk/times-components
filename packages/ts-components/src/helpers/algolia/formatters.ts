import { stripTags } from '@times-components/utils';
import get from 'lodash.get';
import { AlgoliaArticle, IndexedArticle } from './algoliaRelatedArticles';

const formatSummaryContent = (article: IndexedArticle, length: number) => {
  const summary = stripTags(article.content, ' ').trim();
  return summary.substring(0, summary.lastIndexOf(' ', length));
};

const formatSummary = (article: IndexedArticle, length: number) => [
  {
    children: [
      {
        attributes: { value: formatSummaryContent(article, length) },
        children: [],
        name: 'text'
      }
    ],
    name: 'paragraph'
  }
];

const formatImage = (article: AlgoliaArticle, ratio: string) =>
  get(article, 'leadAsset.crop.ratio') === ratio
    ? { url: get(article, 'leadAsset.crop.url') }
    : undefined;

export const formatArticle = (article: any) => ({
  article: {
    id: article.id,
    url: article.url,
    leadAsset: {
      id: get(article, 'leadAsset.id'),
      title: get(article, 'leadAsset.title'),
      crop169: formatImage(article, '16:9')
    },
    label: article.label,
    section: article.section,
    hasVideo: article.hasVideo,
    headline: article.headline,
    shortHeadline: article.shortHeadline,
    bylines: [{ byline: article.byline }],
    publishedTime: article.publishedTime,
    summary125: formatSummary(article, 125),
    summary145: formatSummary(article, 145)
  }
});

type Bylines = Array<{ byline: Array<{ children: any[] }> }>;

export const formatByLines = (bylines: Bylines) =>
  bylines &&
  bylines.length === 1 &&
  bylines[0].byline
    .map(({ children }) =>
      children.map(({ attributes }) => attributes.value).join(' ')
    )
    .join(' ');
