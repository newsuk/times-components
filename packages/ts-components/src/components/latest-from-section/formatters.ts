import { RelatedArticleType } from '../../types/related-article-slice';
type LatestSectionItem = {
  id: string;
  headline: string;
  canonicalID: string;
  summary: string;
  publishedTime: string;
  updatedTime: string;
  author: string;
  slug: string;
  label?: string;
  shortIdentifier: string;
  leadAsset: Array<{
    ratio: string;
    source: string;
  }>;
};

export type LatestSection = {
  section: string;
  items: LatestSectionItem[];
};

export const getSectionTitle = (section: LatestSection) => {
  return section && section.section;
};

export const formatLatestFromSection = (
  section: LatestSection
): RelatedArticleType[] | undefined => {
  return (
    section &&
    section.items.map<RelatedArticleType>((article: LatestSectionItem) => ({
      id: article.id,
      shortIdentifier: article.shortIdentifier,
      slug: article.slug,
      label: article.label || null,
      headline: article.headline,
      section: section.section,
      url: `/article/${article.slug}-${article.shortIdentifier}`,
      leadAsset: {
        crop169: {
          url:
            article.leadAsset &&
            article.leadAsset.find((crop: any) => crop.ratio === '16:9')!.source
        },
        crop32: {
          url:
            article.leadAsset &&
            article.leadAsset.find((crop: any) => crop.ratio === '3:2')!.source
        }
      },
      summary125: [
        {
          name: 'paragraph',
          children: [
            {
              name: 'text',
              attributes: {
                value: article.summary
              },
              children: []
            }
          ]
        }
      ],
      summary145: [
        {
          name: 'paragraph',
          children: [
            {
              name: 'text',
              attributes: {
                value: article.summary
              },
              children: []
            }
          ]
        }
      ],
      publishedTime: article.publishedTime,
      updatedTime: article.updatedTime,
      bylines: [
        {
          byline: [
            {
              name: 'author',
              children: [
                {
                  name: 'text',
                  children: [],
                  attributes: {
                    value: article.author
                  }
                }
              ],
              attributes: {
                slug: 'didi-tang'
              }
            }
          ],
          image: null
        }
      ]
    }))
  );
};
