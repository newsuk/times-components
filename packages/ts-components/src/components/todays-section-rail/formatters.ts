import { RelatedArticleType } from '../../types/related-article-slice';
type TodaysSectionItem = {
  id: string;
  headline: string;
  slug: string;
  url?: string;
  shortIdentifier?: string;
  leadAsset: Array<{
    ratio: string;
    source: string;
  }>;
  media?: {}
};

export type TodaysSection = {
  section: string;
  items: TodaysSectionItem[];
};

export const getSectionTitle = (section: TodaysSection) => {
  return section && section.section;
};

export const formatTodaysSection = (
  section: TodaysSection
): RelatedArticleType[] | undefined => {
  console.log('XXXXXXXXXXXXXXXXXXXXXXXXXXXXX',section)
  return (
    section &&
    section.items.map<RelatedArticleType>((article: TodaysSectionItem) => ({
      id: article.id,
      slug: article.slug,
      label: null,
      shortIdentifier: article.shortIdentifier,
      headline: article.headline,
      section: section.section,
      url: article.url ? article.url :`/article/${article.slug}-${article.shortIdentifier}`,
      leadAsset: {
        crop169: {
          url:
            article.leadAsset &&
            article.leadAsset.find((crop: any) => crop.ratio === '16:9')!.source
        },
        crop32: {
          url:
            article.leadAsset &&
            article.leadAsset.find((crop: any) => crop.ratio === '3:2')!.source
        }
      },
      bylines: []
    }))
  );
};
