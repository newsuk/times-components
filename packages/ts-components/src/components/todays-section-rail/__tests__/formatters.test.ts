import { formatTodaysSection, getSectionTitle } from '../formatters';
import { todaysSection } from '../fixtures/fixtures';
describe('latest from section formatters', () => {
  it('formats', () => {
    expect(formatTodaysSection(todaysSection[4])).toMatchSnapshot();
  });

  it('getSectionTitle', () => {
    expect(getSectionTitle(todaysSection[4])).toEqual('Sport');
  });
});
