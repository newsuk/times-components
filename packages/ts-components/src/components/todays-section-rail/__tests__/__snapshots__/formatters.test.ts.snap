// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`latest from section formatters formats 1`] = `
Array [
  Object {
    "bylines": Array [],
    "headline": "The entire Tour bows to the greatness of Cavendish",
    "id": "c286c868-d90a-11eb-8f14-0bb645f59db0",
    "label": null,
    "leadAsset": Object {
      "crop169": Object {
        "url": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fb7fca746-d8ec-11eb-8f14-0bb645f59db0.jpg",
      },
      "crop32": Object {
        "url": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fb7fca746-d8ec-11eb-8f14-0bb645f59db0.jpg",
      },
    },
    "section": "Sport",
    "shortIdentifier": "2zq0zgfmn",
    "slug": "the-entire-tour-de-france-bows-to-the-greatness-of-mark-cavendish",
    "url": "/article/the-entire-tour-de-france-bows-to-the-greatness-of-mark-cavendish-2zq0zgfmn",
  },
  Object {
    "bylines": Array [],
    "headline": "Flawless Pickford repels everything Germans muster",
    "id": "01ee453e-d90c-11eb-8f14-0bb645f59db0",
    "label": null,
    "leadAsset": Object {
      "crop169": Object {
        "url": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F65e48858-d913-11eb-b92f-5fe539a30c29.jpg",
      },
      "crop32": Object {
        "url": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F65e48858-d913-11eb-b92f-5fe539a30c29.jpg",
      },
    },
    "section": "Sport",
    "shortIdentifier": "ktkcvc85p",
    "slug": "jordan-pickford-repels-everything-germans-muster",
    "url": "/article/jordan-pickford-repels-everything-germans-muster-ktkcvc85p",
  },
  Object {
    "bylines": Array [],
    "headline": "The tactical masterstrokes behind Englandâ€™s ruthless win",
    "id": "a8bfaa82-d917-11eb-b92f-5fe539a30c29",
    "label": null,
    "leadAsset": Object {
      "crop169": Object {
        "url": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F798e1dce-d918-11eb-b92f-5fe539a30c29.jpg",
      },
      "crop32": Object {
        "url": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F65f0ffc0-d918-11eb-b92f-5fe539a30c29.jpg",
      },
    },
    "section": "Sport",
    "shortIdentifier": "s2s5j6k8l",
    "slug": "the-tactical-masterstrokes-behind-englands-ruthless-win",
    "url": "/article/the-tactical-masterstrokes-behind-englands-ruthless-win-s2s5j6k8l",
  },
]
`;
